### PR TITLE
Include DNS response code in DNS trace events

### PR DESCRIPTION
**Pull request by wedaly**
# Include DNS response code in DNS trace events

The DNS response header RCODE field indicates the result of a DNS query, allowing the user to distinguish successful queries from queries that fail with NXDomain or other errors.

## How to use

https://github.com/inspektor-gadget/inspektor-gadget/blob/main/docs/gadgets/trace/dns.md

## Testing done

The local-gadget integration tests have been updated to validate the new field for both success and NXDomain queries. This test implicitly covers the changes to the DNS byte order in PR #1196 as well (without that change, the RCODE field would be FormErr or Refused even for successful responses).

Also tested manually following the steps in https://github.com/inspektor-gadget/inspektor-gadget/blob/main/docs/gadgets/trace/dns.md

image

### CODE REVIEW

File: pkg/gadgets/trace/dns/types/dns.go [hunk 1]
```diff
@@ -35,6 +35,7 @@ type Event struct {
 	PktType    string     `json:"pktType,omitempty" column:"type,minWidth:7,maxWidth:9"`
 	QType      string     `json:"qtype,omitempty" column:"qtype,minWidth:5,maxWidth:10"`
 	DNSName    string     `json:"name,omitempty" column:"name,width:30"`
+	Rcode      string     `json:"rcode,omitEmpty" column:"rcode,minWidth:8"`
```
[the remark this answers is not in the document]
Comment (wedaly) [reply pkg/gadgets/trace/dns/types/dns.go]: thanks! updated

**Comment by wedaly** [comment]
updated TestDNS in local-gadget-manager_test.go, and confirmed it passes with `make local-gadget-tests`

**Comment by mqasimsarfraz** [comment]
Seems like the PR need to be rebased to include the [CI](https://github.com/inspektor-gadget/inspektor-gadget/pull/1220) fix.

**Comment by wedaly** [comment]
> Seems like the PR need to be rebased to include the [CI](https://github.com/inspektor-gadget/inspektor-gadget/pull/1220) fix.

thanks, just rebased

File: pkg/gadgets/trace/dns/tracer/bpf/dns-common.h [hunk 2]
```diff
@@ -22,6 +22,7 @@ struct event_t {
 	// qr says if the dns message is a query (0), or a response (1)
 	unsigned char qr;
 	unsigned char pkt_type;
+	__u8 rcode;
```
Comment (mauriciovasquezbernal) [inline pkg/gadgets/trace/dns/tracer/bpf/dns-common.h:25]: nit: Any reason not to use `unsigned char`? (I know, it's already messed up).

**Review by mauriciovasquezbernal — APPROVED** [review]
Looks good, thanks for the contribution.

File: pkg/gadgets/trace/dns/tracer/bpf/dns-common.h (hunk 2 above)
Comment (wedaly) [reply pkg/gadgets/trace/dns/tracer/bpf/dns-common.h:25]: no particular reason, I'll switch it to `unsigned char` for consistency with the above fields

**Comment by mauriciovasquezbernal** [comment]
Please update https://github.com/inspektor-gadget/inspektor-gadget/blob/17fa2b8d4910294468c2c1b1c5b3146423304b21/integration/inspektor-gadget/integration_test.go#L510 as well.

**Review by mqasimsarfraz — APPROVED** [review]
LGTM!

**Pull request merged**